# OpenVLA：12GB 单卡 QLoRA 近似复现与失败诊断

本 notebook 分别记录 12-episode / 1000-step 小数据实验与 432-episode / 3500-step 全量数据近似实验。只有小数据 adapter 的 episode 12–15 未用于训练；全量 adapter 已见过所有 episode，不能沿用 held-out 标签。

2026-10-01 审计发现旧离线指标存在夹爪编码错误，撤回旧“MAE 改善 19.2%”结论。修正流程与分阶段实验见第 25 节。本机不是论文原配方复现。


In [ ]:
import os

# ==== 全局实验配置 ====
# 本 notebook 从第一步起就固定为「12 个训练 episode + 4 个 held-out episode」，
# 后续训练、adapter smoke test、ground-truth 评测都复用这里的取值。
os.environ['OPENVLA_TRAIN_EPISODES'] = '12'
os.environ['OPENVLA_EVAL_EPISODE_START'] = '12'
os.environ['OPENVLA_EVAL_EPISODES'] = '4'
os.environ['OPENVLA_RUN_ROOT'] = 'runs/libero_spatial_12episode_1000step'
os.environ['OPENVLA_ADAPTER_ROOT'] = 'adapter-tmp/libero_spatial_12episode_1000step'

# ==== 离线运行 ====
os.environ['HF_HUB_OFFLINE'] = '1'
os.environ['TRANSFORMERS_OFFLINE'] = '1'
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

TRAIN_EPISODES = int(os.environ['OPENVLA_TRAIN_EPISODES'])
EVAL_EPISODE_START = int(os.environ['OPENVLA_EVAL_EPISODE_START'])
EVAL_EPISODES = int(os.environ['OPENVLA_EVAL_EPISODES'])
RUN_ROOT = os.environ['OPENVLA_RUN_ROOT']
ADAPTER_ROOT = os.environ['OPENVLA_ADAPTER_ROOT']

print('训练划分    :', f'train[:{TRAIN_EPISODES}]')
print('held-out    :', f'train[{EVAL_EPISODE_START}:{EVAL_EPISODE_START + EVAL_EPISODES}]')
print('run 目录    :', RUN_ROOT)
print('adapter 目录:', ADAPTER_ROOT)


## 1. 硬件设备检测

先确认操作系统、Python、GPU、显存和 CUDA 是否正常。

In [8]:
import os
import platform
import subprocess
import torch

print('操作系统:', platform.system(), platform.release())
print('Python 版本:', platform.python_version())
print('Conda 环境:', os.environ.get('CONDA_DEFAULT_ENV', 'not-set'))
print('CUDA 可用:', torch.cuda.is_available())
print('CUDA 版本:', torch.version.cuda)
print('GPU 数量:', torch.cuda.device_count())
if torch.cuda.is_available():
    print('当前 GPU:', torch.cuda.get_device_name(0))
    print('GPU 显存:', round(torch.cuda.get_device_properties(0).total_memory / 1024**3, 2), 'GB')
try:
    print(subprocess.check_output(['nvidia-smi', '--query-gpu=name,driver_version,memory.total,memory.free', '--format=csv,noheader'], text=True))
except Exception as error:
    print('nvidia-smi 不可用:', error)

操作系统: Linux 6.8.0-138-generic
Python 版本: 3.10.20
Conda 环境: openvla
CUDA 可用: True
CUDA 版本: 12.8
GPU 数量: 1
当前 GPU: NVIDIA GeForce RTX 5070 Ti Laptop GPU
GPU 显存: 11.47 GB
NVIDIA GeForce RTX 5070 Ti Laptop GPU, 595.91.07, 12227 MiB, 2638 MiB



## 2. Python 环境与依赖检查

确认 OpenVLA 所需的主要依赖可以正常导入。

In [9]:
import importlib

required_packages = ['torch', 'transformers', 'peft', 'bitsandbytes', 'PIL', 'numpy', 'pandas', 'ipykernel']
for package_name in required_packages:
    try:
        module = importlib.import_module(package_name)
        print(f'{package_name}: OK -> {getattr(module, "__version__", "unknown")}')
    except Exception as error:
        print(f'{package_name}: MISSING -> {error}')

torch: OK -> 2.8.0+cu128
transformers: OK -> 4.40.1
peft: OK -> 0.11.1
bitsandbytes: OK -> 0.50.2
PIL: OK -> 12.3.0
numpy: OK -> 2.2.5
pandas: OK -> 2.3.3
ipykernel: OK -> 7.3.0


## 3. 仓库状态与关键文件

确认复现入口、训练脚本、adapter 和文档目录都存在。

In [10]:
import os
from pathlib import Path

repo_candidates = []
if os.environ.get('OPENVLA_REPO'):
    repo_candidates.append(Path(os.environ['OPENVLA_REPO']).expanduser())

current_path = Path.cwd().expanduser().resolve()
repo_candidates.extend([current_path, *current_path.parents])

repo = next(
    (
        candidate.resolve()
        for candidate in repo_candidates
        if (candidate / 'reproduction' / 'test_planB_libero_compare.py').exists()
        and (candidate / 'vla-scripts' / 'finetune.py').exists()
    ),
    None,
)
if repo is None:
    raise FileNotFoundError(
        '无法定位 OpenVLA 仓库。请设置 OPENVLA_REPO，或从仓库目录打开 notebook。'
    )

key_paths = [
    'reproduction/run_qlora_1000.py',
    'vla-scripts/finetune.py',
    'reproduction/test_1000step_adapter.py',
    'reproduction/test_planB_libero_compare.py',
    'reproduction/test_planB2_libero_gt.py',
    'adapter-tmp',
    'AIGC_document',
]
for relative_path in key_paths:
    path = repo / relative_path
    print(relative_path, '-> exists' if path.exists() else '-> missing')
print('仓库根目录:', repo)

reproduction/run_qlora_1000.py -> exists
vla-scripts/finetune.py -> exists
reproduction/test_1000step_adapter.py -> exists
reproduction/test_planB_libero_compare.py -> exists
reproduction/test_planB2_libero_gt.py -> exists
adapter-tmp -> exists
AIGC_document -> exists
仓库根目录: /home/yyz/openvla


## 4. 单机兼容性修复

主要修改集中在训练入口，而不是 OpenVLA 基础模型结构。单卡运行时只在分布式进程组已初始化时使用 DDP 和 barrier；保存模型时同时兼容 DDP 包装对象和普通模型对象。

In [11]:
def unwrap_model_for_saving(model):
    return model.module if hasattr(model, 'module') else model

print('单机模式：不包装 DDP，不调用 dist.barrier()')
print('分布式模式：使用 DDP，并通过 model.module 保存原始模型')

单机模式：不包装 DDP，不调用 dist.barrier()
分布式模式：使用 DDP，并通过 model.module 保存原始模型


## 5. 本地缓存与离线运行

使用本地 Hugging Face 缓存，避免 Hub 网络访问影响复现。
本地复现依赖本地模型缓存、离线环境和可用显存。验证 adapter 效果时不需要强制 merge。


In [ ]:
import os
from pathlib import Path

os.environ['HF_HUB_OFFLINE'] = '1'
os.environ['TRANSFORMERS_OFFLINE'] = '1'
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

adapter_root = os.environ.get('OPENVLA_ADAPTER_ROOT', 'adapter-tmp/libero_spatial_12episode_1000step')

hf_home = Path(os.environ.get('HF_HOME', Path.home() / '.cache' / 'huggingface'))
base_cache = hf_home / 'hub' / 'models--openvla--openvla-7b' / 'snapshots'
adapter_dir = repo / adapter_root
print('仓库路径:', repo)
print('base 模型缓存存在:', base_cache.exists())
print('adapter 根目录:', adapter_dir)
print('adapter 目录存在:', adapter_dir.exists())
print('HF_HUB_OFFLINE =', os.environ['HF_HUB_OFFLINE'])
print('TRANSFORMERS_OFFLINE =', os.environ['TRANSFORMERS_OFFLINE'])


## 6. 显存检查

加载 4-bit 模型前，先确认没有其他 Python、Jupyter 或实验进程占用大量显存。

In [13]:
import subprocess

try:
    output = subprocess.check_output(
        ['nvidia-smi', '--query-compute-apps=pid,process_name,used_memory', '--format=csv,noheader'],
        text=True,
    )
    print(output or '当前没有计算进程')
except Exception as error:
    print('无法查看 GPU 进程:', error)

12488, /home/yyz/miniconda3/envs/fast3r/bin/python, 8020 MiB



## 7. 固定 12/4 episode 划分（实验配置）

数据划分不是 notebook 末尾补做的步骤，而是整个实验的前置配置：

- 训练：`train[:12]`（episode 0-11）
- held-out：`train[12:16]`（episode 12-15）

本地 `libero_spatial_no_noops` 共 432 个 episode、16 个 TFRecord 分片（每片 27 个 episode），所以「16 个分片」并不等于「16 个 episode」。这里按 TFDS 的真实 episode 索引划分，保证训练数据与评测数据完全不重叠。

下一节训练通过 `OPENVLA_TRAIN_EPISODES=12` 让 `prismatic/vla/datasets/rlds/dataset.py` 只读取 `train[:12]`；评测脚本通过 `OPENVLA_EVAL_EPISODE_START=12` 与 `OPENVLA_EVAL_EPISODES=4` 只读取 `train[12:16]`。


In [ ]:
import json
import os
from pathlib import Path

train_episodes = int(os.environ.get('OPENVLA_TRAIN_EPISODES', '12'))
eval_start = int(os.environ.get('OPENVLA_EVAL_EPISODE_START', '12'))
eval_episodes = int(os.environ.get('OPENVLA_EVAL_EPISODES', '4'))

repo_candidates = []
if os.environ.get('OPENVLA_REPO'):
    repo_candidates.append(Path(os.environ['OPENVLA_REPO']).expanduser())
current_path = Path.cwd().expanduser().resolve()
repo_candidates.extend([current_path, *current_path.parents])
repo = next(
    (
        candidate.resolve()
        for candidate in repo_candidates
        if (candidate / 'reproduction' / 'test_planB_libero_compare.py').exists()
        and (candidate / 'vla-scripts' / 'finetune.py').exists()
    ),
    None,
)
if repo is None:
    raise FileNotFoundError('无法定位 OpenVLA 仓库，请设置 OPENVLA_REPO。')

manifest = {
    'dataset': 'libero_spatial_no_noops',
    'dataset_total_episodes': 432,
    'dataset_tfrecord_shards': 16,
    'train_split': f'train[:{train_episodes}]',
    'train_episode_range': [0, train_episodes],
    'held_out_split': f'train[{eval_start}:{eval_start + eval_episodes}]',
    'held_out_episode_range': [eval_start, eval_start + eval_episodes],
    'seed': 42,
    'strict_use': 'Retrain on train[:12]; evaluate only on train[12:16].',
}

manifest_path = repo / 'reproduction' / 'held_out_split.json'
manifest_path.write_text(json.dumps(manifest, ensure_ascii=False, indent=2), encoding='utf-8')
print(json.dumps(manifest, ensure_ascii=False, indent=2))
print('清单已保存:', manifest_path)


## 8. 数据划分校验（TFDS episode 切片）

训练前确认两件事：数据集的真实规模，以及 `train[:12]` 与 `train[12:16]` 是否真的落在不同的 episode 上。校验通过后再启动训练，避免把评测数据混进训练集。


In [ ]:
import os
from pathlib import Path

import tensorflow_datasets as tfds

train_episodes = int(os.environ.get('OPENVLA_TRAIN_EPISODES', '12'))
eval_start = int(os.environ.get('OPENVLA_EVAL_EPISODE_START', '12'))
eval_episodes = int(os.environ.get('OPENVLA_EVAL_EPISODES', '4'))

dataset_root = Path(os.environ.get('LIBERO_RLDS_ROOT', Path.home() / 'modified_libero_rlds'))
dataset_name = 'libero_spatial_no_noops'

builder = tfds.builder(dataset_name, data_dir=str(dataset_root))
split_info = builder.info.splits['train']
print('数据集        :', dataset_name)
print('总 episode    :', split_info.num_examples)
print('TFRecord 分片 :', split_info.num_shards)
print('每片 episode  :', list(split_info.shard_lengths))
print('是否存在 val  :', 'val' in builder.info.splits)

train_split = f'train[:{train_episodes}]'
held_out_split = f'train[{eval_start}:{eval_start + eval_episodes}]'

for split in (train_split, held_out_split):
    dataset = tfds.load(dataset_name, data_dir=str(dataset_root), split=split, shuffle_files=False)
    episode_count = 0
    first_instruction = None
    for example in dataset:
        if first_instruction is None:
            first_step = next(iter(example['steps'].as_numpy_iterator()))
            first_instruction = first_step['language_instruction'].decode('utf-8', errors='replace')
        episode_count += 1
    print(f'{split}: {episode_count} 个 episode | 首条指令: {first_instruction}')


## 9. LoRA 微调（启动脚本已参数化）

`reproduction/run_qlora_1000.py` 现在支持用参数选择数据规模、步数和输出目录，默认值仍是 12-episode 实验：

```bash
# 12-episode 实验（第 1 节里的默认配置）
python reproduction/run_qlora_1000.py

# 论文口径：官方全量 LIBERO-Spatial（432 条 demo）+ 约 1 个 epoch
python reproduction/run_qlora_1000.py \
  --train-episodes 432 --max-steps 3500 --save-steps 500 \
  --run-name libero_spatial_full_paper
```

超参与官方一致：4-bit NF4 量化 + LoRA rank 32 / all-linear + lr 5e-4 + batch size 2 × 梯度累积 8（等效 batch 16）+ image_aug。
脚本会自动用 `systemd-inhibit --what=idle:sleep:handle-lid-switch` 阻止空闲休眠与合盖休眠——笔记本一旦 suspend，CUDA 上下文会失效、训练进程会卡死（已实际发生过一次）。

输出路径规则：`runs/<run-name>/<exp_id>/` 与 `adapter-tmp/<run-name>/<exp_id>/`，日志为 `reproduction/finetune_<run-name>.log`。

各次训练的实际结果见第 15 / 18 节。


In [ ]:
import os
import subprocess
import sys

RUN_TRAINING = False  # 改成 True 才会启动 12-episode / 1000-step 长任务

environment = os.environ.copy()
environment['OPENVLA_TRAIN_EPISODES'] = str(int(os.environ.get('OPENVLA_TRAIN_EPISODES', '12')))
environment['OPENVLA_EVAL_EPISODE_START'] = str(int(os.environ.get('OPENVLA_EVAL_EPISODE_START', '12')))
environment['OPENVLA_EVAL_EPISODES'] = str(int(os.environ.get('OPENVLA_EVAL_EPISODES', '4')))

if RUN_TRAINING:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'run_qlora_1000.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout[-4000:])
    if result.stderr:
        print(result.stderr[-2000:])
else:
    print('安全模式：跳过长时间训练；需要时将 RUN_TRAINING 改为 True。')


## 10. adapter smoke test

执行 12-episode / 1000-step adapter 的最小推理验证，确认 4-bit base 模型、LoRA adapter 和推理流程都能工作。

（下面的输出是上一版全量数据 adapter 的记录，重新训练后执行本节会覆盖为新的 adapter 结果。）


In [14]:
import os
import subprocess
import sys

environment = os.environ.copy()
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

RUN_ADAPTER_SMOKE = False  # 改成 True 才会加载 7B + LoRA 并运行推理
if RUN_ADAPTER_SMOKE:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'test_1000step_adapter.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout)
    if result.stderr:
        print(result.stderr)
else:
    print('安全模式：跳过 adapter smoke test；需要时将 RUN_ADAPTER_SMOKE 改为 True。')

returncode: 1
OpenVLA 1000-step QLoRA adapter smoke test

[1/5] Loading processor...
Processor loaded.

[2/5] Loading 4-bit base OpenVLA...

I0000 00:00:1789882248.821267   21472 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
I0000 00:00:1789882249.137304   21472 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1789882250.147328   21472 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environme

## 11. LIBERO 对比验证（旧输出撤回）

旧脚本 generate 漏传 pixel_values，旧输出不能证明视觉条件下的微调效果。已补齐图像输入，但单张普通图片仍不足以衡量策略能力。生产推理路径诊断见第 25 节。


In [ ]:
import os
import subprocess
import sys

environment = os.environ.copy()
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

RUN_COMPARE = False  # 改成 True 才会运行 base/LoRA 动作对比
if RUN_COMPARE:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'test_planB_libero_compare.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print('CSV exists:', (repo / 'planB_libero_compare.csv').exists())
    print(result.stdout[-4000:])
    if result.stderr:
        print(result.stderr[-2000:])
else:
    print('安全模式：跳过 base/LoRA 动作对比；需要时将 RUN_COMPARE 改为 True。')

## 12. LIBERO ground-truth 对比（历史指标失效）

旧评测将 policy 夹爪 0=闭合/1=张开与 TFDS 原始 +1=闭合/-1=张开直接相减。旧 MAE/MSE/L2、相对变化与逐样本胜率均不可引用。旧脚本已修正 GT 夹爪，但只作兼容入口；新版诊断调用生产 get_vla_action，分别报告六维运动误差、夹爪正确率、归一化 L1 和 teacher-forcing 指标。

只有 12-episode adapter 的 train[12:16] 可称未用于训练；原归一化统计来自全量数据，因此不是严格完全隔离的验证。


In [ ]:
import os
import subprocess
import sys

environment = os.environ.copy()
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

RUN_GROUND_TRUTH = False  # 改成 True 才会重新生成 12/4 held-out CSV
if RUN_GROUND_TRUTH:
    result = subprocess.run(
        [sys.executable, str(repo / 'reproduction' / 'test_planB2_libero_gt.py')],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout[-5000:])
    if result.stderr:
        print(result.stderr[-3000:])
else:
    print('安全模式：跳过 held-out ground-truth 评测；需要时将 RUN_GROUND_TRUTH 改为 True。')

## 13. 读取修正后的指标

不再读取没有版本标记的旧 CSV。新 JSON 未产生时明确显示 pending。


In [ ]:
import json
result_path = repo / 'reproduction/results/small1000_heldout_v1.json'
if result_path.exists():
    corrected = json.loads(result_path.read_text())
    assert corrected['schema_version'] == 1 and corrected['held_out_from_training']
    print('样本数:', corrected['samples'], '| 数据:', corrected['split'])
    print(json.dumps(corrected['summary'], indent=2, ensure_ascii=False))
else:
    print('修正后的 held-out 指标 pending；旧 CSV 指标已撤回。')


## 14. 多 checkpoint 对比

固定评测流程，只改变 checkpoint：

- base model
- 12-episode / 1000-step adapter
- 后续更长训练得到的 adapter

每个 checkpoint 都使用同一批 held-out 样本、指令、图像和评价指标，这样才能判断训练步数是否带来稳定改善。


In [20]:
checkpoint_root = repo / 'adapter-tmp'
checkpoints = sorted(checkpoint_root.glob('**/adapter_config.json'))
if checkpoints:
    print('发现的 adapter checkpoint:')
    for checkpoint in checkpoints:
        print(checkpoint.parent)
else:
    print('暂未发现 adapter checkpoint。完成更多训练后，再将 checkpoint 加入统一评测。')

发现的 adapter checkpoint:
/home/yyz/openvla/adapter-tmp/47a0ec7fc4ec123775a391911046cf33cf9ed83f+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug
/home/yyz/openvla/adapter-tmp/libero_spatial_1000step/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug


## 15. 实验结果（12 训练 / 4 held-out 协议）

**训练**（日志：`reproduction/finetune_12episode_1000step.log`）

- 数据：`train[:12]`，即 432 个 episode 里的前 12 个（16 个 TFRecord 分片、每片 27 个 episode）
- 配置：4-bit NF4 + LoRA rank 32 / all-linear + lr 5e-4 + batch size 2 + 梯度累积 8 + 1000 steps
- 耗时：2026-09-21 07:36 → 10:30，1000/1000 步跑完（约 2 小时 54 分，末段 10.3 s/it）
- adapter：`adapter-tmp/libero_spatial_12episode_1000step/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug/`（968 MB）
- 训练结束后不再合并成完整权重（`merge_lora_during_training=False`）：合并需要额外约 15GB 内存和 15GB 磁盘，而所有评测脚本都直接加载 LoRA adapter
- 训练期间用 `systemd-inhibit --what=idle:sleep` 阻止空闲休眠：笔记本一旦 suspend，CUDA 上下文会失效，训练进程会卡死不再前进

**旧 held-out 指标撤回（2026-10-01）**

原指标表、19.2% 改善和逐样本胜率依赖错误夹爪编码，不能引用或写进简历。样本未参与训练不能挽救错误指标。修正数值待重测。


In [ ]:
import json
result_path = repo / 'reproduction/results/small1000_heldout_v1.json'
if result_path.exists():
    corrected = json.loads(result_path.read_text())
    assert corrected['schema_version'] == 1 and corrected['held_out_from_training']
    print('样本数:', corrected['samples'], '| 数据:', corrected['split'])
    print(json.dumps(corrected['summary'], indent=2, ensure_ascii=False))
else:
    print('修正后的 held-out 指标 pending；旧 CSV 指标已撤回。')


## 16. LIBERO simulator 成功率评测

动作误差只是间接指标，最终指标是 LIBERO simulator 里的任务成功率。本机第一次使用 LIBERO，需要先做一次环境准备：

```bash
# 1) 克隆并安装 LIBERO（setup.py 的 install_requires 为空，不会动 openvla 环境里的 torch）
git clone https://github.com/Lifelong-Robot-Learning/LIBERO.git ~/LIBERO
cd ~/LIBERO && pip install -e .

# 2) 对齐 mujoco 版本：mujoco>=3.2 的 pybind11 枚举会让 robosuite 1.4.1 里的
#    `joint_type in (mjJNT_HINGE, mjJNT_SLIDE)` 断言失败，固定到 3.1.6
pip install "mujoco==3.1.6"

# 3) 让 import libero 生效（editable 安装只映射了 libero.libero 子包）
echo "$HOME/LIBERO" > "$(python -c 'import site; print(site.getsitepackages()[0])')/libero_path.pth"

# 4) 非交互生成 LIBERO 路径配置（默认会在 import 时弹交互式提问）
#    ~/.libero/config.yaml 内容见仓库说明
```

另外做了三处兼容性修改（原因都写在代码注释里）：

- `~/LIBERO/libero/libero/benchmark/__init__.py`：`torch.load(..., weights_only=False)`，否则 torch>=2.6 无法加载 LIBERO 的 init states
- `experiments/robot/openvla_utils.py`：支持 LoRA adapter 目录（base 模型 + PeftModel）、processor 回退到 base 模型、用 `sdpa` 替代本机没有的 flash-attn、`norm_stats` 同时写到内部 base 模型
- `experiments/robot/libero/run_libero_eval.py`：新增 `--base_model_path`，用于指定 LoRA 对应的 base 模型

评测命令（冒烟用 `num_trials_per_task 2`，正式对齐论文用 50，即 10 个任务 × 50 次 = 500 个 rollout）：

```bash
MUJOCO_GL=egl python experiments/robot/libero/run_libero_eval.py \
  --model_family openvla \
  --pretrained_checkpoint <ADAPTER_DIR> \
  --task_suite_name libero_spatial \
  --num_trials_per_task 50 \
  --load_in_4bit True
```

结果写入 `experiments/logs/EVAL-libero_spatial-openvla-<时间>.txt`（每个任务的成功次数 + 总成功率）。


In [ ]:
import os
import subprocess
import sys

# 论文口径：10 任务 × 50 次 = 500 个 rollout；冒烟时先用 2
NUM_TRIALS_PER_TASK = 50
RUN_LIBERO_EVAL = False  # 改成 True 才会启动 500-rollout 长任务

# 选择要评测的 adapter：12-episode 实验 / 论文口径全量实验
RUN_NAME = os.environ.get('OPENVLA_RUN_NAME', 'libero_spatial_full_paper')

adapter_dir = repo / 'adapter-tmp' / RUN_NAME / (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
if not (adapter_dir / 'adapter_config.json').is_file():
    raise FileNotFoundError(f'adapter 不存在: {adapter_dir}')

environment = os.environ.copy()
environment['MUJOCO_GL'] = 'egl'                     # 离屏渲染走 EGL
environment['HF_HUB_OFFLINE'] = '1'
environment['TRANSFORMERS_OFFLINE'] = '1'
environment['TOKENIZERS_PARALLELISM'] = 'false'

if RUN_LIBERO_EVAL:
    result = subprocess.run(
        [
            sys.executable,
            str(repo / 'experiments/robot/libero/run_libero_eval.py'),
            '--model_family', 'openvla',
            '--pretrained_checkpoint', str(adapter_dir),
            '--base_model_path', 'openvla/openvla-7b',
            '--task_suite_name', 'libero_spatial',
            '--num_trials_per_task', str(NUM_TRIALS_PER_TASK),
            '--center_crop', 'True',
            '--load_in_4bit', 'True',
            '--seed', '7',
            '--run_id_note', RUN_NAME,
        ],
        cwd=repo,
        env=environment,
        text=True,
        capture_output=True,
        check=False,
    )
    print('returncode:', result.returncode)
    print(result.stdout[-5000:])
    if result.stderr:
        print(result.stderr[-2000:])
else:
    print('安全模式：跳过 500-rollout LIBERO 评测；需要时将 RUN_LIBERO_EVAL 改为 True。')


## 17. LIBERO simulator 成功率结果（12-episode / 1000-step adapter）

评测设置：`libero_spatial` 全部 10 个任务 × 10 次 rollout = 100 个 episode，4-bit 量化加载 base + LoRA adapter，
`MUJOCO_GL=egl` 离屏渲染。结果日志：`experiments/logs/EVAL-libero_spatial-openvla-2026_09_21-18_53_03--12ep_1000step_10trials.txt`。

**总成功率：28.0%（28/100）**

| 任务 | 训练数据里出现过？ | 成功率 |
|---|---|---|
| 1. bowl between the plate and the ramekin | 是（1 条 episode） | **90%** |
| 2. bowl next to the ramekin | 否 | 0% |
| 3. bowl from table center | 否 | 0% |
| 4. bowl on the cookie box | 否 | 0% |
| 5. bowl in the top drawer of the wooden cabinet | 是（3 条） | 10% |
| 6. bowl on the ramekin | 是（4 条） | **60%** |
| 7. bowl next to the cookie box | 是（1 条） | 30% |
| 8. bowl on the stove | 是（3 条） | **90%** |
| 9. bowl next to the plate | 否 | 0% |
| 10. bowl on the wooden cabinet | 否 | 0% |

关键结论：**训练时出现过的 5 个任务合计 28/50 = 56%，没出现过的 5 个任务合计 0/50 = 0%。**
上述结果与任务覆盖相关，但没有隔离训练时长、量化与随机性，不能认定覆盖范围是唯一或主要根因。全量数据模型后来在 382 个 rollout 仅成功 5 次，说明“只扩大数据就会改善”没有得到本项目验证。

清洗后的本地数据是 432 个 episode；原始演示数量不等于实际训练样本量。后续先诊断训练帧动作与短闭环，不仅为增加评测次数继续耗时。


In [ ]:
from pathlib import Path

import pandas as pd
import tensorflow_datasets as tfds

# 1) 训练数据覆盖了哪些任务（train[:12] 的指令分布）
dataset_root = str(Path.home() / 'modified_libero_rlds')
train_split = tfds.load(
    'libero_spatial_no_noops',
    data_dir=dataset_root,
    split='train[:12]',
    shuffle_files=False,
)
train_coverage = {}
for example in train_split:
    step = next(iter(example['steps'].as_numpy_iterator()))
    instruction = step['language_instruction'].decode('utf-8', errors='replace')
    train_coverage[instruction] = train_coverage.get(instruction, 0) + 1

# 2) 解析最新的 LIBERO 评测日志
log_candidates = sorted((repo / 'experiments/logs').glob('EVAL-libero_spatial-*10trials*.txt'))
if not log_candidates:
    raise FileNotFoundError('没有找到 LIBERO 评测日志，请先执行第 16 节。')
log_path = log_candidates[-1]
print('评测日志:', log_path.name)

rows, current_task = [], None
for line in log_path.read_text(encoding='utf-8', errors='replace').splitlines():
    if line.startswith('Task: '):
        current_task = line[len('Task: '):].strip()
    elif line.startswith('Current task success rate:') and current_task:
        rows.append({'task': current_task, 'success_rate': float(line.split(':')[1])})
        current_task = None

table = pd.DataFrame(rows)
table['train_demos'] = table['task'].map(lambda t: train_coverage.get(t, 0))
table['success_rate_%'] = (table['success_rate'] * 100).round(1)
table = table[['task', 'train_demos', 'success_rate_%']]
print(table.to_string(index=False))

seen = table[table['train_demos'] > 0]
unseen = table[table['train_demos'] == 0]
print()
print(f"训练覆盖到的任务: {len(seen)} 个 | 平均成功率 {seen['success_rate_%'].mean():.1f}%")
print(f"训练未覆盖的任务: {len(unseen)} 个 | 平均成功率 {unseen['success_rate_%'].mean() if len(unseen) else float('nan'):.1f}%")
print(f"整体成功率: {table['success_rate_%'].mean():.1f}%")


## 18. 论文对齐实验（全量数据 + 500 次 rollout）

对照 OpenVLA 论文 Appendix E 与仓库 README 的 LIBERO 设置，本项目与论文的对应关系：

| 项目 | 论文 | 本项目 |
|---|---|---|
| 训练数据 | 官方清洗后的 LIBERO-Spatial：500 条演示中剔除 68 条失败演示 = **432 条** | `train[:432]`，同一份 `libero_spatial_no_noops` RLDS 数据 |
| 数据清洗 | 256×256 重渲染、删除 no-op 动作、第三人称图像旋转 180°、剔除失败演示 | 全部由官方 RLDS 提供（数据集名里的 `no_noops` 即第 2 项） |
| 微调方式 | LoRA r=32，作用于所有线性层；官方 checkpoint 不使用量化 | LoRA r=32、all-linear，但本地为 4-bit NF4 QLoRA 近似配置 |
| 数据增强 | 每帧随机裁剪 90% 面积 | `--image_aug true`，评测端 `center_crop` 取中心 90% |
| 评测 | 每个 suite 500 次 rollout（10 任务 × 50 次），3 个随机种子取平均 | `--num_trials_per_task 50`（1 个种子，3 个种子约需 3 倍时间） |
| 参考成绩 | LIBERO-Spatial **84.7 ± 0.9%**（论文的 3 seeds 平均） | 仅作参考，不能当作本机结果 |

需要特别区分“全量数据”和“官方训练配方”：全量 52,970 个 transition ÷ 16 ≈ 3,300 步只代表约 1 个 epoch，并不等于官方训练已经完成。官方 [LIBERO-Spatial model card](https://huggingface.co/openvla/openvla-7b-finetuned-libero-spatial) 明确记录 **50K gradient steps、batch size 128、训练/测试均不量化、无 gradient accumulation、8×A100**；本机的 3,500 steps、有效 batch 16、4-bit NF4 和单卡环境只是资源受限的近似实验。
本地全量训练实测约 11.5 小时，但它只覆盖官方 50K-step 配方的很小一部分，因此不能预期直接达到论文成绩。

运行记录：训练已经完成；正式 simulator 评测需要单独运行并通过下一节的完整性检查。

- 本地近似训练：`python reproduction/run_qlora_1000.py --train-episodes 432 --max-steps 3500 --save-steps 500 --run-name libero_spatial_full_paper`
- 官方参考配方：50,000 gradient steps、batch size 128、无量化；在当前 12GB GPU 上不应直接宣称已经复现。
- 每 500 步保存一次 adapter（长跑中断也能留下可用权重）
- 评测：第 20 节，`NUM_TRIALS_PER_TASK = 50`，`RUN_NAME = 'libero_spatial_full_paper'`

> **方法学边界**：本实验使用 4-bit NF4、只跑 1 个 seed，和论文的 3 seeds × 500 rollouts 并不完全相同。更准确的说法是“在本地资源约束下完成了论文设置的近似复现”，而不是声称复现了论文的 84.7 ± 0.9%。

## 19. 复现状态审计：训练完成 ≠ 评测完成

这一节只读取本地日志和权重，不会重新训练，也不会把计划中的数字当成结果。全量模型使用 `train[:432]`，因此它已经见过 `train[12:16]`；第 13 节的 12/4 held-out 动作误差只能用于 **12-episode 实验**，不能作为全量模型的 held-out 结果。

我们把项目拆成两条独立证据链：

1. **12-episode 研究实验**：`train[:12]` → `train[12:16]`，已有 MAE/MSE/L2 和 100-rollout LIBERO 结果。
2. **全量论文对齐实验**：`train[:432]` → `libero_spatial` 500 rollouts；训练权重已保存，成功率要以正式评测日志为准。


In [ ]:
from pathlib import Path
import re
import pandas as pd

full_run_name = 'libero_spatial_full_paper'
full_exp_id = (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
full_adapter = repo / 'adapter-tmp' / full_run_name / full_exp_id
full_train_log = repo / 'reproduction' / 'finetune_libero_spatial_full_paper.log'
eval_log_candidates = sorted((repo / 'experiments' / 'logs').glob(
    'EVAL-libero_spatial-openvla-*full_paper*50trials*.txt'
))

def parse_eval_log(path):
    text = path.read_text(encoding='utf-8', errors='replace') if path.exists() else ''
    task_rates = []
    for match in re.finditer(
        r'Task: (?P<task>.+?)\n.*?Current task success rate: (?P<rate>[0-9.]+)',
        text,
        flags=re.S,
    ):
        task_rates.append({'task': match.group('task').strip(), 'success_rate': float(match.group('rate'))})
    episode_counts = [int(value) for value in re.findall(r'# episodes completed so far: (\d+)', text)]
    success_counts = [int(value) for value in re.findall(r'# successes: (\d+) \(', text)]
    return {
        'path': path,
        'bytes': path.stat().st_size if path.exists() else 0,
        'task_rates': task_rates,
        'episodes': max(episode_counts, default=0),
        'successes': max(success_counts, default=0),
    }

train_text = full_train_log.read_text(encoding='utf-8', errors='replace') if full_train_log.exists() else ''
training_complete = (
    '100%|██████████| 3500/3500' in train_text
    and 'Max step 3500 reached! Stopping training...' in train_text
    and (full_adapter / 'adapter_config.json').is_file()
    and (full_adapter / 'adapter_model.safetensors').is_file()
)
eval_records = [parse_eval_log(path) for path in eval_log_candidates]
best_eval = max(eval_records, key=lambda item: (item['episodes'], item['bytes']), default=None)
formal_eval_complete = bool(
    best_eval
    and best_eval['episodes'] >= 500
    and len(best_eval['task_rates']) >= 10
)
formal_eval_interrupted = bool(best_eval and not formal_eval_complete and best_eval['episodes'] > 0)

status = pd.DataFrame([
    {'stage': '12-episode held-out action metrics', 'status': 'legacy invalid; corrected pending', 'evidence': str(repo / 'planB2_libero_gt.csv')},
    {'stage': '12-episode LIBERO simulator (10 × 10)', 'status': 'complete', 'evidence': '28/100 = 28.0% in 2026-09-21 log'},
    {'stage': 'full-data QLoRA training (432 episodes, 3500 steps)', 'status': 'complete' if training_complete else 'incomplete', 'evidence': str(full_train_log)},
    {'stage': 'full-data LIBERO simulator (10 × 50)', 'status': 'complete' if formal_eval_complete else ('interrupted' if formal_eval_interrupted else 'pending'), 'evidence': str(best_eval['path']) if best_eval else 'no usable 50-trial log'},
])
print(status.to_string(index=False))
print()
print('full adapter:', full_adapter)
print('adapter size (decimal MB):', round((full_adapter / 'adapter_model.safetensors').stat().st_size / 10**6, 1) if (full_adapter / 'adapter_model.safetensors').exists() else 'missing')
if best_eval:
    print('best full-paper eval log:', best_eval['path'].name)
    print('episodes parsed:', best_eval['episodes'], '| task summaries:', len(best_eval['task_rates']))
else:
    print('尚未找到可用于汇报的 full-paper simulator 评测日志。')


## 20. Full-paper LIBERO 评测：可复现运行单元（默认安全模式）

全量 adapter 已经存在，不需要重新训练。正式评测是 10 个任务 × 50 次 rollout，预计需要数小时；因此下面的 `RUN_FORMAL_EVAL` 默认是 `False`，先打印命令和检查项。确认 GPU、MuJoCo 和 LIBERO 都正常后，把它改为 `True` 再执行。

评测中断后不要读取半截日志当结果；重新启动完整 500-rollout 评测，并用第 21 节检查是否真的出现 10 个 task summary 和 500 个 episode。

In [ ]:
import os
import subprocess
import sys

RUN_FORMAL_EVAL = False  # 改成 True 才会启动长时间评测
NUM_TRIALS_PER_TASK = 50
adapter_dir = repo / 'adapter-tmp' / 'libero_spatial_full_paper' / (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
if not (adapter_dir / 'adapter_config.json').is_file():
    raise FileNotFoundError(f'full adapter 不存在: {adapter_dir}')

eval_command = [
    sys.executable, str(repo / 'experiments/robot/libero/run_libero_eval.py'),
    '--model_family', 'openvla',
    '--pretrained_checkpoint', str(adapter_dir),
    '--base_model_path', 'openvla/openvla-7b',
    '--task_suite_name', 'libero_spatial',
    '--num_trials_per_task', str(NUM_TRIALS_PER_TASK),
    '--center_crop', 'True',
    '--load_in_4bit', 'True',
    '--seed', '7',
    '--run_id_note', 'full_paper_50trials',
]
eval_env = os.environ.copy()
eval_env.update({
    'MUJOCO_GL': 'egl',
    'HF_HUB_OFFLINE': '1',
    'TRANSFORMERS_OFFLINE': '1',
    'TOKENIZERS_PARALLELISM': 'false',
})
print('adapter:', adapter_dir)
print('command:', ' '.join(eval_command))
gpu_probe = subprocess.run(['nvidia-smi', '--query-gpu=name,memory.free', '--format=csv,noheader'], text=True, capture_output=True)
print('GPU probe:', gpu_probe.stdout.strip() or gpu_probe.stderr.strip())
if RUN_FORMAL_EVAL:
    completed = subprocess.run(eval_command, cwd=repo, env=eval_env, text=True, check=False)
    print('评测退出码:', completed.returncode)
else:
    print('安全模式：没有启动长时间评测。确认资源可用后，将 RUN_FORMAL_EVAL 改为 True。')


## 21. 正式评测结果回填（只接受完整日志）

运行第 20 节后重新执行本节。代码会选择已解析 episode 数最多的 full-paper 日志，并只在 `500 episodes + 10 个 task summary` 都存在时计算总成功率。这样 notebook 不会把旧 adapter、冒烟结果或中途退出的日志混在一起。

In [ ]:
if not eval_records:
    print('没有 full-paper 评测日志；当前正式成功率仍为 pending。')
elif not formal_eval_complete:
    print(f'日志 {best_eval["path"].name} 只解析到 {best_eval["episodes"]}/500 个 episode、{len(best_eval["task_rates"])} 个 task summary。')
    print('不要把这个中间状态写成最终 success rate。')
else:
    task_table = pd.DataFrame(best_eval['task_rates'])
    task_table['success_rate_%'] = (task_table['success_rate'] * 100).round(1)
    print(task_table[['task', 'success_rate_%']].to_string(index=False))
    total_success_rate = best_eval['successes'] / best_eval['episodes']
    print()
    print(f'Full-paper success rate: {best_eval["successes"]}/{best_eval["episodes"]} = {total_success_rate:.1%}')
    print(f'日志: {best_eval["path"]}')


## 22. 结论与可诚实汇报的范围

- 12GB 单卡 QLoRA 训练工程跑通：1000-step 小数据和 3500-step 全量权重已保存，可注明近似实验。
- 旧 held-out 动作误差改善：编码错误，已撤回，不可引用。
- 小数据 simulator 结果：28/100，只描述该次实验，不推断唯一根因。
- 全量部分评测：停止时 5/382 ≈ 1.3%，可用于失败诊断，不是完整 500-rollout 结果。
- 新离线诊断及 10-update pilot：只引用第 25 节实际 JSON；尚未生成则 pending。

论文参考指标不是本机结果；目前没有达到论文成功率的证据。


## 23. 简历友好版项目总结（更正旧指标）

> **OpenVLA-7B：12GB 单卡 QLoRA 近似复现与失败诊断**
> - 实现 NF4 + LoRA 训练和 LIBERO-Spatial 仿真评测，完成 1000-step 小数据与 432-episode / 3500-step 全量数据近似实验。
> - 审计并修复图像输入遗漏、夹爪编码不一致和梯度累积保存节点问题，构建动作空间单元测试、teacher-forcing / 自回归诊断及闭环动作追踪。
> - 用 notebook 管理配置、日志和证据，报告 28/100 小数据结果与全量模型 5/382 的中断评测，区分已确认错误、待验证原因和论文参考值。

不要写旧“MAE 改善 19.2%”，不要声称论文成绩复现成功。工程修复与失败诊断可以成为贡献，但要能够解释实验边界。


## 24. 低成功率排查：先停止长评测，再定位误差来源

本节记录一次有意中断的 full-paper 评测和后续诊断。评测在 **382/500 episodes、5 次成功** 时停止，临时累计成功率为 **1.3%**；它不是最终 500-rollout 指标，不能替代完整实验结果。已经保留原始日志和 rollout 视频。

### 证据链

1. **不是 adapter 没加载。** 运行时类型为 `PeftModel`，内部模型为 `OpenVLAForActionPrediction`，`norm_stats` 含 `libero_spatial_no_noops`，adapter 文件约 969 MB，LoRA 权重张量非零。
2. **不是 simulator 崩溃。** 382 个 episode 都能完成并保存视频，没有渲染、文件句柄或环境异常；因此低分是策略行为问题，不是评测程序提前报错。
3. **小探针只排除部分明显异常，未定位根因。** 用 12 个真实 RLDS 帧做的短探针中，按相同 `center_crop`、`predict_action`、数据统计量反归一化流程得到的物理动作 MAE 约为 0.061。这个小样本不是正式指标，不足以证明完整部署接口正确，不能当作根因已确认的证据。
4. **训练明显没有达到论文 recipe。** 本地日志末尾的 `action_accuracy` 约 0.420、`l1_loss` 约 0.111；本地只训练了 3,500 steps、有效 batch 16、4-bit NF4、单卡。官方 LIBERO-Spatial checkpoint 的 model card 记录的是 50K gradient steps、batch 128、训练和测试均不量化、无 gradient accumulation、8×A100。

### 目前最可能的错误位置（按优先级）

- **第一优先级：训练配方/训练量不一致。** 3,500 steps 约等于本地数据的一轮，而官方是 50,000 steps；按有效 batch 估算，本地看到约 56,000 个样本，官方配方约 6.4M 个样本，数量级相差约 114 倍。同时 QLoRA、batch 16 和单卡与官方无量化 batch 128 的 checkpoint 不同。这可能造成学习不足，但需要受控实验验证，不能把当前 1.3% 解释成论文模型的能力。
- **第二优先级：数据到闭环的分布差异。** no-op 删除、单一初始状态/策略、多样性不足会让 teacher-forcing 动作误差看起来尚可，但 rollout 中一旦偏离演示轨迹就无法恢复。
- **第三优先级：图像/动作接口。** 当前已核对 224×224 processor、90% center crop、`libero_spatial_no_noops` 统计键和 gripper 归一化方向；仍需在任何新 checkpoint 上做 1 个 task × 10 rollout 的小实验。
- **第四优先级：软件/硬件差异。** 本机是 PyTorch 2.8 + bitsandbytes 0.50 + SDPA；官方结果使用 PyTorch 2.2、flash-attn 2.5.5、A100。它是复现误差来源，但目前证据弱于训练量和量化差异。

路线已确定为“12GB 单卡 QLoRA 近似复现 + 失败诊断”。下一节先纠正指标、诊断训练帧，再短程验证训练/保存链路。有证据才扩大训练，不自动重启 500 个 rollout。


In [ ]:
from pathlib import Path
import json
import re

diag_adapter = repo / 'adapter-tmp' / 'libero_spatial_full_paper' / (
    'openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+'
    'lora-r32+dropout-0.0+q-4bit--image_aug'
)
diag_stats = diag_adapter / 'dataset_statistics.json'
diag_log = repo / 'experiments' / 'logs' / 'EVAL-libero_spatial-openvla-2026_09_29-23_21_29--full_paper_50trials_retry_20260929.txt'
assert (diag_adapter / 'adapter_config.json').is_file()
assert (diag_adapter / 'adapter_model.safetensors').is_file()
adapter_cfg = json.loads((diag_adapter / 'adapter_config.json').read_text())
stats = json.loads(diag_stats.read_text())
text = diag_log.read_text(encoding='utf-8', errors='replace') if diag_log.exists() else ''
episodes = max([int(x) for x in re.findall(r'# episodes completed so far: (\d+)', text)] or [0])
successes = max([int(x) for x in re.findall(r'# successes: (\d+) \(', text)] or [0])
task_rates = re.findall(r'Current task success rate: ([0-9.]+)', text)
print('base model:', adapter_cfg['base_model_name_or_path'])
print('norm stats keys:', list(stats))
print('adapter size MB:', round((diag_adapter / 'adapter_model.safetensors').stat().st_size / 10**6, 1))
print('interim episodes/successes:', episodes, successes)
print('interim success rate:', f'{successes / episodes:.1%}' if episodes else 'n/a')
print('completed task summaries:', len(task_rates), task_rates)
print('formal result accepted:', episodes >= 500 and len(task_rates) >= 10)
print('short action probe (manual, not formal): physical MAE ≈ 0.061 on 12 RLDS frames')


## 25. 12GB 单卡 QLoRA：纠错与分阶段继续（2026-10-01）

### 已确认的问题与未证实的推断

| 位置 | 证据 | 影响与边界 |
|---|---|---|
| 旧离线 GT 评测 | policy gripper 0/1 与 raw gripper +1/-1 直接相减 | 旧误差结论失效；不在 simulator 路径上，不能解释其 1.3% |
| 旧视觉对比 | generate 未传 pixel_values | 不能证明视觉条件下的微调效果；已修复 |
| 梯度累积循环 | 按微批次保存、最终更新后额外 backward | 重复保存、记录不清；已修复，尚无导致低分的证据 |
| 训练规模与量化 | 本地 3500 × 16 / NF4；官方 50000 × 128 / 无量化 | 学习不足和量化损失是高优先级假设，不是已验证因果结论 |
| 训练/评测 double quant | 训练 False，旧部署 True | 新诊断显式使用 False；单帧无差别不足以排除全局影响 |
| 失败动作形态 | 部分视频有停滞、抓起后未正确放下 | 怀疑运动精度、偏离演示后的恢复或释放时机；需动作 trace 核验 |

### 实验安排

1. `diagnose_qlora.py`：每个 episode 固定起始、中间、末段非终止帧，比较 base / adapter / 不裁剪 adapter，并对相同裁剪画面计算 teacher-forcing 指标。保存动作、图像 hash、动作空间说明和 GPU 峰值。全量模型的训练帧诊断不是泛化测评，也不是任务成功率。
2. 在小数据 adapter 的 episode 12–15 重测 corrected held-out 指标。固定 12 帧，不是旧随机 50 帧，二者不可当作同一实验比较。原归一化统计来自全量数据，因此不是严格完全隔离的验证流程。
3. 从 3500-step adapter **权重热启动**：新目录、新优化器、lr=1e-4，10 次优化更新（80 微批次）。这是训练/保存链路验证，不是模型改进证据，不是精确断点续训；核对 `training_state.json`。
4. 同一帧集合重测 pilot。完成后审核，再设计学习率/时长的受控实验，不自动启动长训练。10 步指标的波动不能证明因果关系或策略改善。
5. 闭环入口支持 `--task_ids '[0]' --num_trials_per_task 2 --trace_actions True --fail_fast True --bnb_double_quant False`：记录每步动作、末端位置和夹爪状态，异常立即报告。默认完整评测协议不变。先核验两个固定初始状态，失败后按停滞、抓取、搬运、释放阶段定位，再决定是否扩大到每任务少量 rollout。

### 如何读诊断

- 训练帧 teacher-forcing 与自回归都差：优先检查学习量、目标动作/图像配对、量化加载和权重。
- teacher-forcing 好但自回归差：优先查解码/提示格式、前序动作 token 依赖和误差累积；不直接等同于训练不足。
- 离线动作好但闭环差：优先查图像/相机与仿真接口、动作频率、精确放置和偏离演示轨迹后的恢复。
- 不裁剪和裁剪差异显著：提示图像预处理敏感，但少量帧不足以证明应该关闭裁剪；需闭环受控对照。

运行阶段 1：

```bash
# 默认只打印计划，不占 GPU、不训练
python reproduction/run_diagnostic_stage.py
# Fast3R 结束、GPU 空闲后才运行；不会终止其他 GPU 进程
python reproduction/run_diagnostic_stage.py --execute
```

CPU 回归测试：`python -m unittest reproduction.test_action_metrics -v`，5 项通过（夹爪往返、旧指标反例、运动裁剪、量化夹爪端点与反归一化）。首次 GPU 诊断在加载时 OOM，Fast3R 当时占约 5.7GB 显存；这不是策略失败证据。按用户要求保留 Fast3R，等待完成。结果未产生时明确标记 pending。

官方配方：[LIBERO-Spatial model card](https://huggingface.co/openvla/openvla-7b-finetuned-libero-spatial)。官方建议核对训练画面上的部署预测及演示回放：[performance troubleshooting](https://github.com/openvla/openvla#vla-performance-troubleshooting)。


In [1]:
import json
for filename in ('full3500_diagnostic_v1.json', 'small1000_heldout_v1.json', 'warmstart10_diagnostic_v1.json'):
    path = repo / 'reproduction/results' / filename
    if not path.exists():
        print(filename, 'pending（未执行成功，不填数字）')
        continue
    report = json.loads(path.read_text())
    assert report['schema_version'] == 1 and report['samples'] > 0
    print(filename, '| samples:', report['samples'], '| held-out:', report['held_out_from_training'])
    print('GPU peak GiB:', report['peak_allocated_gib'])
    print(json.dumps(report['summary'], indent=2, ensure_ascii=False))


full3500_diagnostic_v1.json | samples: 24 | held-out: False
GPU peak GiB: 5.710696220397949
{
  "base": {
    "motion_mae_6d": 0.20562120203210243,
    "simulator_mae_7d": 0.23577055412275447,
    "normalized_l1_7d": 0.3854217533798608,
    "gripper_correct": 0.7916666666666666,
    "translation_norm": 0.5633954202047099,
    "normalized_mae_per_dim": [
      0.3685971025414983,
      0.30978569933511196,
      0.5166343597480733,
      0.43120663575311574,
      0.5100348648441015,
      0.35156289248287637,
      0.21013071895424837
    ]
  },
  "adapter": {
    "motion_mae_6d": 0.0851838588972254,
    "simulator_mae_7d": 0.07301473619762178,
    "normalized_l1_7d": 0.16362727785713524,
    "gripper_correct": 1.0,
    "translation_norm": 0.2960704442029036,
    "normalized_mae_per_dim": [
      0.12050329901325407,
      0.13444921765319554,
      0.21167902556036852,
      0.2024850045924691,
      0.2920190245897225,
      0.18164099450596938,
      0.002614379084967311
    ]
  },


## 26. 第一轮修正诊断的实际结果（2026-10-01）

| 模型/数据 | 样本 | 六维运动 MAE | 归一化七维 L1 | 夹爪正确率 |
|---|---:|---:|---:|---:|
| 全量 adapter：训练帧 | 24 | 0.085184 | 0.163627 | 100.0% |
| 12-episode adapter：未参与训练帧 | 12 | 0.149595 | 0.267444 | 75.0% |

全量 adapter 的 teacher-forcing token accuracy = 36.31%，归一化 token 解码 L1 = 0.117180。自由生成的连续目标 L1 = 0.163627，不裁剪时 = 0.202607。前者使用量化 token 目标、后者使用连续归一化目标，不能把差值全算作 exposure bias。

这一小样本显示 adapter 确实改善了动作预测、夹爪整体方向没有写反；teacher-forcing L1 与历史训练日志约 0.111 同量级，但尚未达到高精度动作拟合。旋转维度仍有明显误差。这不证明 1.3% 的唯一根因；需要官方 checkpoint 对照和闭环追踪。

本地与官方 checkpoint 的 action q01/q99/mask 已对照：q01 最大差约 5.4e-9、q99 与 mask 完全一致。因此当前未发现反归一化范围明显用错的证据。

结果只来自固定起始/中间/末段帧，不能代表整体样本分布或任务成功率。小数据结果是新 12 帧实验，不恢复旧随机 50 帧的 19.2% 结论。GPU 峰值 allocated 约 5.71GiB，不是系统总显存或训练峰值。

### 后续状态

- 全量本地 adapter 的 task0 × 2 固定初始状态闭环诊断已完成，NF4、double quant=False、center crop=True、seed=7；实际行为见第27节。
- 官方 checkpoint 固定 revision `962318cec55ac10993ff0f5f43eda9a270b4c873`，14个文件共15,085,047,523 bytes，已完成下载及逐文件大小核验，记录于 `results/official_checkpoint_manifest.json`；下载成功不等于对照评测成功。
- 完整 checkpoint 和 adapter 控制组统一显式使用 NF4，避免完整模型的裸 load_in_4bit 参数默认采用 FP4。此修复用于控制实验一致性，不是原低分已确认根因。
- 10-update 热启动暂缓：官方权重下载优先，保存新 adapter 前保留至少 2GiB 磁盘余量，不允许填满系统盘。


## 27. 闭环失败定位：起始状态停滞（不是抓取后放置失败）

同一个 task0 的两个固定初始状态均未成功（0/2），没有运行异常，每个执行220个动作。这只是小规模诊断，不代表整个suite成功率。

| 诊断量 | 初始状态0 | 初始状态1 |
|---|---:|---:|
| 平移命令范数<0.02的比例 | 100.00 | 99.55 |
| 末端离起点最大距离（毫米） | 2.942 | 3.651 |
| 张开夹爪命令比例 | 100.0 | 100.0 |

机器人几乎没有接近物体，夹爪也始终保持张开，因此这两个失败属于起始停滞，而非抓起后释放失败。两种失败模式可能同时存在，但必须按实际trace区分。

训练帧起始状态也出现此模式：8个起始帧中7个预测相同近零平移，GT平移并不全为零。这是具体行为证据，不只是泛泛的“QLoRA性能下降”。Fast3R结束后，官方同NF4配置、同task/初始状态对照已完成，见第28节。

原始HDF5演示和完整sim初始化状态在本机未找到，演示回放仍待完成；RLDS末端状态不等同于MuJoCo完整状态，不将不匹配的回放当作数据验证。


## 28. 官方 NF4 对照与本地失败诊断（2026-10-01）

固定官方revision `962318cec55ac10993ff0f5f43eda9a270b4c873`；不训练、不重下载。两组均NF4、BF16 compute、double quant=False、center crop=True。24张图像hash、episode/step、指令和原始动作目标一致。这仍是训练分布内的诊断，不是泛化测评。

| 指标 | 本地3500-step adapter | 官方完整checkpoint，NF4部署 |
|---|---:|---:|
| 六维运动MAE | 0.085184 | 0.013161 |
| 连续归一化七维L1 | 0.163627 | 0.028775 |
| 夹爪方向正确数 | 24/24 | 24/24 |
| teacher-forcing token accuracy | 36.31% | 94.05% |
| teacher-forcing decoded-token L1 | 0.117180 | 0.006956 |

官方不裁剪连续L1=0.087103；少量帧不能证明全局预处理因果关系。官方峰值PyTorch allocated=4.918GiB，计算28.08秒（不含加载），不是训练峰值。

闭环都是task0、trial0/1、seed7、220动作上限及相同等待步数；首次策略查询的末端和夹爪qpos逐值相同（未独立hash整个MuJoCo状态）。

| 模型 / 状态 | 成功 | 动作数 | 平移命令范数<0.02比例 | 末端最大位移 | 夹爪命令切换 |
|---|---|---:|---:|---:|---:|
| 本地 / 0 | 否 | 220 | 100.00% | 2.94mm | 0 |
| 官方NF4 / 0 | 是 | 80 | 3.75% | 383.40mm | 2 |
| 本地 / 1 | 否 | 220 | 99.55% | 3.65mm | 0 |
| 官方NF4 / 1 | 否 | 220 | 88.64% | 15.01mm | 0 |

官方1/2、本地0/2是两个固定状态的诊断计数，不能写成整套LIBERO的50%/0%成功率，也不能声称统计显著性。两组均无运行异常。

推断与边界：官方同NF4在状态0成功，加上更好的训练帧拟合，不支持把本地极低表现全归因于4-bit部署或共同仿真链路。重点转向本地训练产物、监督与优化流程；但状态1官方也停滞，共同接口/图像/状态敏感性没有完全排除。没有官方BF16对照，不能测量量化损失；NF4推理对照也不是QLoRA训练无损的证明。

后续未完成：原始HDF5演示回放、状态1图像/初始化诊断，以及有足够空间后的10-update保存链路验证。系统盘仅约2GiB空余，暂不训练或扩大500rollout。此轮有界对照完成，完整高成功率复现尚未完成。简历可如实描述“12GB单卡QLoRA近似复现与失败诊断”，不得写成已达到论文性能。

下一单元只读取真实结果并核验可比性，不加载GPU模型。


In [1]:
import json
from pathlib import Path
result_root = Path('reproduction/results') if Path('reproduction/results').exists() else Path('results')
local = json.loads((result_root / 'full3500_diagnostic_v1.json').read_text())
official = json.loads((result_root / 'official_diagnostic_v1.json').read_text())
assert local['samples'] == official['samples'] == 24
assert local['quantization'] == official['quantization']
frame_fields = ('episode', 't', 'image_sha256', 'instruction', 'raw_target')
assert all(all(a[k] == b[k] for k in frame_fields) for a, b in zip(local['rows'], official['rows']))
print('24 fixed images, instructions, targets and quantization: matched')
for name, report, key in [('local', local, 'adapter'), ('official', official, 'official')]:
    metrics = report['summary'][key]
    print(name, 'motion MAE:', round(metrics['motion_mae_6d'], 6), 'normalized L1:', round(metrics['normalized_l1_7d'], 6))
    print(name, 'teacher token accuracy:', round(report['summary']['teacher_forced']['token_accuracy'], 6))
metadata_fields = ('task_id', 'instruction', 'trial', 'seed', 'center_crop', 'double_quant')
for trial in (0, 1):
    a = json.loads(next((result_root / 'sim_local_full3500').glob(f'*trial{trial}.json')).read_text())
    b = json.loads(next((result_root / 'sim_official').glob(f'*trial{trial}.json')).read_text())
    assert all(a[k] == b[k] for k in metadata_fields)
    assert a['actions'][0]['eef_pos'] == b['actions'][0]['eef_pos']
    assert a['actions'][0]['gripper_qpos'] == b['actions'][0]['gripper_qpos']
    assert a['error'] is None and b['error'] is None
    print('trial', trial, 'metadata and first observable state: matched; successes:', a['success'], b['success'])
for filename in ('local_full3500_task0_trace_summary.json', 'official_task0_trace_summary.json'):
    report = json.loads((result_root / filename).read_text())
    assert report['episodes'] == 2
    print(filename, f"{report['successes']}/{report['episodes']}", '(fixed-state diagnostic only)')


24 fixed images, instructions, targets and quantization: matched
local motion MAE: 0.085184 normalized L1: 0.163627
local teacher token accuracy: 0.363095
official motion MAE: 0.013161 normalized L1: 0.028775
official teacher token accuracy: 0.940476
trial 0 metadata and first observable state: matched; successes: False True
trial 1 metadata and first observable state: matched; successes: False False
local_full3500_task0_trace_summary.json 0/2 (fixed-state diagnostic only)
official_task0_trace_summary.json 1/2 (fixed-state diagnostic only)


## 29. 训练链路体检：CPU阶段快照（2026-10-01；后续GPU结果见第30节）

真实结果：`results/training_chain_cpu_audit_v1.json`，可重复脚本：`reproduction/audit_training_chain.py`。此轮不加载GPU模型、不做参数更新、不覆盖原权重。

| 检查 | 实际范围与结果 |
|---|---|
| 图像 / 指令 / 动作配对 | 前8条演示951帧，经实际RLDS标准化、goal relabeling、窗口和resize后与同TFDS step对齐 |
| 动作归一化 / 夹爪 | BOUNDS_Q99最大绝对差1.83e-7，夹爪保留1=open / 0=close |
| 动作监督 | 24固定帧各7个动作token+1个EOS，动作token与独立离散化一致 |
| 标签padding / 提示屏蔽 | collator检查通过；未截断监督标签，padding不参与loss |
| 训练 / 推理输入 | 相同图像pixel_values完全一致；补29871空白token后提示前缀相同 |
| checkpoint | 879张量均finite；878个LoRA张量+完整lm_head，LoRA参数110,828,288 |
| 完整lm_head | PEFT自动保存；FP32权重与base逐值一致 |

范围限制：禁用随机增强以对齐；不是历史随机增强核验，也不是原始HDF5 → TFDS转换验证。951帧中有5帧平移范数<0.02，含8个terminal帧，不能外推全量空闲动作比例或认定terminal标签一定无效。

439个LoRA B矩阵中426个非零；13个全零位于DINO最后block23、SigLIP最后block26及attention pool。当前模型返回倒数第二层patch，这些分支不影响该输出，因此不能把它们当作有效分支未训练的证据。非零权重也不能证明优化过程正确。导出adapter配置的inference_mode=True不等于训练时被冻结；热启动显式is_trainable=True。

GPU仍被Fast3R占用约8.4GiB，不抢占。**尚未执行**：真实动作loss反传的LoRA梯度检查、一次optimizer update参数差、保存/重新加载预测一致性。旧`test_qlora_backward.py`用input_ids作为假labels，不是这些检查的替代品。

结论：CPU可验证部分未发现新的明显链路错误，但第一步尚未全部完成，不能宣布低成功率根因已找到。下面单元只读取真实报告，不启动训练。


In [1]:
import json
from pathlib import Path
root = Path('reproduction/results') if Path('reproduction/results').exists() else Path('results')
audit = json.loads((root / 'training_chain_cpu_audit_v1.json').read_text())
assert audit['schema_version'] == 1
assert all(v == 'pass' for v in audit['checks'].values())
print(json.dumps(audit['counts'], indent=2))
print('Max normalization error:', audit['max_normalization_abs_difference'])
print('LoRA parameters:', audit['lora_parameter_count'])
print('Saved lm_head equals base:', audit['saved_lm_head_equals_base_fp32'])
b_tensors = [t for t in audit['checkpoint_tensors'] if 'lora_B' in t['key']]
print('Nonzero LoRA B tensors:', sum(t['nonzero'] for t in b_tensors), '/', len(b_tensors))
prior = json.loads((root / 'full3500_diagnostic_v1.json').read_text())
assert [(x['episode'], x['t'], x['image_sha256']) for x in audit['rows']] == [(x['episode'], x['t'], x['image_sha256']) for x in prior['rows']]
print('24 prior diagnostic frames matched')
print('Pending:', audit['pending'])


{
  "episodes": 8,
  "paired_frames": 951,
  "supervised_frames": 24,
  "supervised_action_tokens": 168,
  "supervised_eos_tokens": 24,
  "terminal_frames": 8,
  "raw_small_translation_frames": 5
}
Max normalization error: 1.8305317195377313e-07
LoRA parameters: 110828288
Saved lm_head equals base: True
Nonzero LoRA B tensors: 426 / 439
24 prior diagnostic frames matched
Pending: ['runtime_LoRA_gradients', 'actual_optimizer_update', 'save_reload_prediction_equivalence', 'original_HDF5_image_action_alignment', 'historical_random_augmentation_samples']


## 30. 运行时体检完成：一次更新与保存重载通过（2026-10-01）

真实结果 `results/training_chain_runtime_audit_v1.json`：status=pass。Fast3R释放GPU后才执行，没有抢占。第29节和CPU JSON保留当时的pending快照；其中前三项runtime检查现已补齐，HDF5来源配对和历史随机增强仍待验证。

执行范围仅16个已核验固定帧、microbatch2×累积8、AdamW lr5e-4、1次优化更新。记录真实动作loss、每个LoRA张量梯度是否finite/非零、更新前后参数差；禁用随机增强，此工程检查不代表性能改进或精确断点续训。

独立临时adapter采用与训练相同的默认PEFT保存行为，退出后清理；原始adapter/config/stats前后SHA256须一致。保存前检查磁盘可用空间至少为原adapter大小+1GiB保留空间+64MiB辅助缓冲。GPU忙或磁盘不足直接拒绝，不删除用户数据。

8个微批次loss均finite，2.262401–3.528154，平均2.819472。878个LoRA张量中852个有finite、非零梯度，且852个发生参数变化，最大绝对差0.000501126；冻结参数均无梯度。其余26个A/B张量恰对应13个不参与patch输出的视觉末层/池化分支，不是有效分支全被冻结。

更新后3个阶段帧中2个动作改变，不代表动作改善。相同训练加载器重载和实际生产加载器重载，都与保存前更新后的3×7动作逐值一致，最大生产动作差0。训练准备模型用BF16 autocast，生产入口保持无autocast，未混淆两种上下文；动作一致不等于内部logits/dtype或全部输入一致。

训练阶段PyTorch峰值allocated 7.957GiB；临时保存971,136,009 bytes，退出后已清理，原权重/config/stats SHA256不变，没有保留新checkpoint或优化器。首次尝试在更新前因体检预测缺失autocast发生FP32/BF16冲突，已保留 `results/training_chain_runtime_attempt1_error.json` 并修复后有界重试；不是历史低成功率根因。

第一步当前训练链路体检完成：样本上不支持LoRA完全没梯度、optimizer不更新或保存丢失adapter的解释；不证明历史训练全过程正确、充分收敛、QLoRA无性能损失或论文成功率已复现。建议下一轮设计有界少样本拟合并同时看teacher-forcing和自回归动作，本轮不自动启动。

```bash
# 默认dry-run；结果已存在时 --execute 拒绝覆盖，不重复更新
python reproduction/audit_training_runtime.py \
  --checkpoint 'adapter-tmp/libero_spatial_full_paper/openvla-7b+libero_spatial_no_noops+b16+lr-0.0005+lora-r32+dropout-0.0+q-4bit--image_aug'
```


In [1]:
import json, math, statistics
from pathlib import Path
root = Path('reproduction/results') if Path('reproduction/results').exists() else Path('results')
runtime = json.loads((root / 'training_chain_runtime_audit_v1.json').read_text())
cpu = json.loads((root / 'training_chain_cpu_audit_v1.json').read_text())
assert runtime['status'] == 'pass' and runtime['optimizer_updates'] == 1
assert runtime['samples'] == 16 and runtime['batch_size'] * runtime['accumulation_steps'] == 16
assert all(math.isfinite(x) for x in runtime['losses'])
assert len(runtime['gradients']) == len(runtime['parameter_changes']) == 878
inactive = {x['key'].replace('.lora_B.weight', '.lora_B.default.weight') for x in cpu['checkpoint_tensors'] if 'lora_B' in x['key'] and not x['nonzero']}
inactive |= {x.replace('.lora_B.', '.lora_A.') for x in list(inactive)}
assert {x['name'] for x in runtime['gradients'] if not x['grad_present']} == inactive
assert {x['name'] for x in runtime['parameter_changes'] if x['max_abs_change'] == 0} == inactive
assert all(math.isfinite(x['grad_norm']) and x['grad_nonzero'] for x in runtime['gradients'] if x['grad_present'])
assert runtime['after_update_actions'] == runtime['same_training_loader_reloaded_actions'] == runtime['production_loader_reloaded_actions']
assert runtime['production_max_abs_action_difference'] == 0
assert not runtime['original_checkpoint_modified'] and runtime['temporary_checkpoint_cleaned']
print('Runtime:', runtime['status'], '| real frames:', runtime['samples'], '| optimizer updates:', runtime['optimizer_updates'])
print('Finite nonzero gradient / changed tensors:', runtime['gradient_tensors_nonzero'], '/', runtime['changed_parameter_tensors'])
print('26 inactive vision-branch tensors: names matched CPU evidence')
print('Mean real-action loss:', round(statistics.mean(runtime['losses']), 6))
print('Same-training and production reload: exact 3-frame action match')
print('Original checkpoint unchanged; temporary checkpoint cleaned')
print('Limits: one update, no augmentation, no policy improvement claim')


Runtime: pass | real frames: 16 | optimizer updates: 1
Finite nonzero gradient / changed tensors: 852 / 852
26 inactive vision-branch tensors: names matched CPU evidence
Mean real-action loss: 2.819472
Same-training and production reload: exact 3-frame action match
Original checkpoint unchanged; temporary checkpoint cleaned
Limits: one update, no augmentation, no policy improvement claim


## 31. 有界拟合启动与第10步快照（2026-10-02；最终结果见第32节）

系统盘扩容后检查可用约122GB；用户授权继续，宿主GPU空闲后已启动独立本地worker。`fit_small_sample.py` 从3500-step adapter热启动，前16个已核验真实帧，batch2×累积8，lr1e-4，最多50更新；无随机增强，无中心裁剪。不是历史训练的精确续训，不是泛化/成功率证明。

0/10/25/50节点评估同24帧teacher-forcing和自回归动作；另外8帧不用于本轮拟合，但原全量adapter已经见过，不能标held-out。诊断图像不裁剪，不能与之前中心裁剪的指标当作同一实验比较；lr1e-4不是学习率消融。生产加载器在结束时独立复测，差异不能隐藏。

启动时已通过14项CPU回归和AST检查，当时最终报告尚待产生；此段保留历史快照，现已完成的最终证据见第32节及下面只读单元。CPU测试本身不代表GPU拟合成功。

第10步中间证据已核验：`results/small_fit_milestone10_v1.json`。拟合16帧token准确率38.39%→97.32%，自回归运动MAE 0.112703→0.005097，归一化L1 0.214299→0.006506，夹爪15/16→16/16，近零平移9→1；probe8帧token准确率仍32.14%，运动MAE 0.100045→0.092145。当时50步和生产重载尚未结束；不可用中间节点替代最终结果或泛化证明。

`runs/small_fit_v1` 保留原子progress、独立日志和0/10/25/50不可覆盖checkpoint，包含adapter、optimizer和Torch CPU/CUDA RNG、SHA256 manifest；完整目录发布后才可恢复，不使用partial。原权重保留，更新数上限50，GPU忙或磁盘不足拒绝，重复worker锁避免双开。

本阶段曾创建每小时heartbeat，收尾时关闭，不自动扩大下一实验。独立Python计算不调用Codex/API，但机器须不关机/休眠；定时接手需桌面应用运行和可用额度，不保证额度刷新瞬间恢复，不使用免费重置券或付费API。接手记录见 `reproduction/CONTINUE.md`。


In [1]:
import json
from pathlib import Path
project = Path.cwd() if Path('reproduction/results').exists() else Path.cwd().parent
interim = json.loads((project / 'reproduction/results/small_fit_milestone10_v1.json').read_text())
assert interim['status'] == 'interim_checkpoint10_verified_not_final'
assert [e['step'] for e in interim['evaluations']] == [0, 10]
print('Verified interim: 0/10 checkpoints; not final fitting or success rate')
for e in interim['evaluations']:
    print('fit step', e['step'], json.dumps(e['summary']['fit']))
report_path = project / 'reproduction/results/small_fit_v1.json'
progress_path = project / 'runs/small_fit_v1/progress.json'
if report_path.exists():
    report = json.loads(report_path.read_text())
    assert report['optimizer_updates'] == 50
    assert [x['step'] for x in report['evaluations']] == [0, 10, 25, 50]
    assert not report['original_checkpoint_modified']
    print('Fitting report:', report['status'])
    for evaluation in report['evaluations']:
        print('step', evaluation['step'], json.dumps(evaluation['summary']))
    print('Production actions equal:', report['production_actions_equal'])
    verification = json.loads((project / 'reproduction/results/small_fit_verification_v1.json').read_text())
    import hashlib
    assert hashlib.sha256(report_path.read_bytes()).hexdigest() == verification['result_sha256']
    assert verification['status'] == 'pass' and all(v == 'pass' for v in verification['checks'].values())
    print('Verified pre-update mean training losses:', verification['mean_training_loss_before_update'])
    print('Probe teacher-token L1 by loader:', verification['probe_teacher_token_l1_at_step50'])
    print('Limits:', report['limits'])
elif progress_path.exists():
    progress = json.loads(progress_path.read_text())
    print('Worker status:', progress['status'])
    print('In-memory updates:', progress['completed_updates_in_memory'])
    print('Completed checkpoint:', progress['latest_completed_checkpoint'])
    print('Final metrics: pending; no rollout or generalization claim')
else:
    print('Local worker evidence unavailable; final metrics pending')


Verified interim: 0/10 checkpoints; not final fitting or success rate
fit step 0 {"samples": 16, "teacher_token_accuracy": 0.3839285714285714, "teacher_token_l1": 0.14593837535014006, "autoregressive_motion_mae": 0.11270311026229714, "autoregressive_normalized_l1": 0.21429853684243833, "gripper_correct": 15, "small_translation_count": 9}
fit step 10 {"samples": 16, "teacher_token_accuracy": 0.9732142857142858, "teacher_token_l1": 0.002240896358543418, "autoregressive_motion_mae": 0.005096859865909591, "autoregressive_normalized_l1": 0.006505965335235883, "gripper_correct": 16, "small_translation_count": 1}
Fitting report: completed
step 0 {"fit": {"samples": 16, "teacher_token_accuracy": 0.3839285714285714, "teacher_token_l1": 0.14593837535014006, "autoregressive_motion_mae": 0.11270311026229714, "autoregressive_normalized_l1": 0.21429853684243833, "gripper_correct": 15, "small_translation_count": 9}, "probe_previously_seen": {"samples": 8, "teacher_token_accuracy": 0.3214285714285714,

## 32. 50步小样本拟合完成：能学习动作，不等于策略泛化

真实报告 `results/small_fit_v1.json` 状态completed；只读核验 `results/small_fit_verification_v1.json`。不重跑GPU，四节点完整snapshot spec/hash、24帧身份、原始目标、逐帧动作指标和汇总均核验。50条loss各含8微批次且finite，源adapter/config/stats SHA256未变。

| 更新后节点 | fit16 token准确率 | fit运动MAE | fit归一化L1 | fit近零平移数 | probe8 token准确率 | probe运动MAE | probe归一化L1 |
|---|---:|---:|---:|---:|---:|---:|---:|
| 0 | 38.39% | 0.112703 | 0.214299 | 9 | 32.14% | 0.100045 | 0.166197 |
| 10 | 97.32% | 0.005097 | 0.006506 | 1 | 32.14% | 0.092145 | 0.162852 |
| 25 | 100.00% | 0.001350 | 0.002039 | 0 | 32.14% | 0.087087 | 0.170275 |
| 50 | 100.00% | 0.001350 | 0.002039 | 0 | 32.14% | 0.091907 | 0.177316 |

运动MAE只计算六维连续运动；归一化L1以连续训练目标作参考；token准确率是teacher-forcing，不是自回归或任务成功率。近零阈值为平移命令范数<0.02，不是末端实际位移。fit25/50虽动作token全对，连续MAE仍非零，不能把量化token中心当连续GT。fit夹爪最终16/16；probe从8/8降至7/8。

平均训练loss在更新1/10/25/50之前分别为2.819472、0.283406、0.000280078、0.0000210585；表格动作测量在更新之后。未单独测第0步loss，不把更新1标为第0步。训练峰值PyTorch allocated 9.217GiB，不是系统总显存。

生产loader（无autocast）重载最终checkpoint后，24×7自回归动作与训练loader（BF16 autocast）逐值一致。但probe teacher-token L1训练0.127591、生产0.124930，不能声称loader全部等价；fit两者准确率100%、token L1为0。保留真实差异，不把dtype/logits混成全PASS。

这证明当前QLoRA/真实动作监督/优化组合能记住16帧，减弱“4bit配置根本学不会动作”的解释；8个probe无对应改善且后期L1变差，不能宣称泛化或闭环提升。probe曾被原adapter训练见过，**不是held-out**。历史失败根因仍未唯一确定，lr1e-4、新AdamW、禁用增强和无中心裁剪的影响尚未分离；不能由小样本拟合推断充分训练或论文成功率。

四个独立checkpoint、optimizer/Torch RNG和日志保留且不提交大文件；核验了snapshot完整性，但未实际测试中断GPU恢复，不承诺位级复现。14项CPU测试、AST、Notebook schema和下方reader输出再核验。接手自动化在本阶段收尾时关闭，不启动长训练/500评测。

建议下一步仅做用户确认的单因素有界对照：保持同16帧、同源adapter及所有设置，lr改为历史5e-4，在独立run比较同节点；不能直接代替历史训练复原。真正泛化需episode级独立训练/验证划分，不能用本轮probe替代。


## 33. 学习率对照启动与第10步快照（最终结果见第34节）

用户2026-10-02授权继续。空闲GPU和约46GB可用磁盘确认后，独立worker运行 `fit_small_sample.py --experiment small_fit_lr5e4_v1 --launch`。从与lr1e-4完全相同的3500-step adapter开始，不从上轮step050继续。

同16拟合真实帧、8个原adapter见过的probe，batch2×累积8、新AdamW、seed7、NF4/BF16/double quant=False、无随机增强/无中心裁剪；仅lr改为5e-4，最多50更新。spec除lr外完全相同，第0步24帧全部输出与参考逐值匹配、snapshot和源hash已核验，记录 `results/small_fit_lr5e4_start_v1.json`。不匹配即在首次更新前停止。

启动时最终报告目标 `results/small_fit_lr5e4_v1.json`；比较目标 `results/small_fit_lr_comparison_v1.json`。现已产生并核验，见第34节。下面reader不训练、不推理，其存储输出已更新到真实最终状态；若文件缺失则明确标pending，不补造指标。

22项CPU回归通过（包括对照配置/起点一致、真实摘要/图像身份篡改和生产loader标识检测）；AST与Notebook schema另行核验。CPU通过不是GPU对照已完成。报告应区分teacher-forcing与自回归、生产loader差异；loss在更新前，动作表在更新后。两组各一个seed/run，不能宣称显著性或历史唯一根因。8probe不是held-out，无simulator rollout或成功率结论。

第10步中间证据 `results/small_fit_lr5e4_milestone10_v1.json` 已核验snapshot hash、frame及原始目标。lr5e-4 fit token100%、运动MAE0.001350；参考lr1e-4同节点97.32%、0.005097。probe分别33.93%/0.070397与32.14%/0.092145。这个节点较大学习率更快拟合，不支持它在本轮小样本中必然失效，但不能替代最终50步对照或历史全量训练结论。

原权重与前轮结果保留；独立run含四节点可恢复snapshot、optimizer/Torch RNG和日志，不提交重量文件。本阶段曾创建每小时heartbeat，完成后关闭；接手记录 `CONTINUE.md` 保留，不自动启动长训练/500评测。


In [1]:
import json, sys
from pathlib import Path
project = Path.cwd() if Path('reproduction/results').exists() else Path.cwd().parent
if str(project) not in sys.path:
    sys.path.insert(0, str(project))
start = json.loads((project / 'reproduction/results/small_fit_lr5e4_start_v1.json').read_text())
assert start['status'] == 'verified_start_not_final'
assert start['step0_rows_exactly_match_reference'] and start['original_hashes_verified']
print('Verified matched step0; only configuration change: lr1e-4 to lr5e-4')
final = project / 'reproduction/results/small_fit_lr5e4_v1.json'
if final.exists():
    from reproduction.compare_small_fit_lr import compare
    comparison = compare(False)
    print('Final reports available; portable row/spec checks:', comparison['status'])
    for node in comparison['milestones']:
        print('step', node['step'], json.dumps(node))
    print('Production loaders:', json.dumps(comparison['production']))
    evidence = project / 'reproduction/results/small_fit_lr_comparison_v1.json'
    if evidence.exists():
        verified = json.loads(evidence.read_text())
        assert verified['reports_sha256'] == comparison['reports_sha256']
        assert verified['status'] == 'verified_comparison'
        assert verified['milestones'] == comparison['milestones'] and verified['production'] == comparison['production']
        print('Local snapshot/source verification:', verified['local_snapshot_verification'])
        for lr, checks in verified['local_checks'].items():
            assert checks['snapshot_and_source_hashes'] == 'pass'
            print('Pre-update mean losses', lr, checks['mean_loss_before_update'])
    else:
        print('Full local snapshot/loss verification still pending')
else:
    progress = project / 'runs/small_fit_lr5e4_v1/progress.json'
    if progress.exists():
        state = json.loads(progress.read_text())
        print('Worker status:', state['status'])
        print('In-memory updates:', state['completed_updates_in_memory'])
        print('Completed snapshot:', state['latest_completed_checkpoint'])
        if 'error' in state:
            print('Error:', state['error'])
    print('Final LR comparison pending; no generalization or success-rate claim')


Verified matched step0; only configuration change: lr1e-4 to lr5e-4
Final reports available; portable row/spec checks: verified_comparison
step 0 {"step": 0, "lr1e-4": {"fit": {"samples": 16, "teacher_token_accuracy": 0.3839285714285714, "teacher_token_l1": 0.14593837535014006, "autoregressive_motion_mae": 0.11270311026229714, "autoregressive_normalized_l1": 0.21429853684243833, "gripper_correct": 15, "small_translation_count": 9}, "probe_previously_seen": {"samples": 8, "teacher_token_accuracy": 0.3214285714285714, "teacher_token_l1": 0.11218487394957984, "autoregressive_motion_mae": 0.10004491293099038, "autoregressive_normalized_l1": 0.1661974397850028, "gripper_correct": 8, "small_translation_count": 6}}, "lr5e-4": {"fit": {"samples": 16, "teacher_token_accuracy": 0.3839285714285714, "teacher_token_l1": 0.14593837535014006, "autoregressive_motion_mae": 0.11270311026229714, "autoregressive_normalized_l1": 0.21429853684243833, "gripper_correct": 15, "small_translation_count": 9}, "pr

## 34. 学习率有界对照完成：拟合更快，迁移指标并非全面更好

完整真实报告 `results/small_fit_lr5e4_v1.json` completed；只读比较 `results/small_fit_lr_comparison_v1.json` verified_comparison、local_snapshot_verification=True。两组源hash与配置除lr外一致，起始24帧全部结果相同。8个完整snapshot保护文件hash、原源adapter/config/stats、24帧身份/真实目标及逐帧动作指标/汇总独立重算通过；各50条loss、每条8微批次finite。未重跑GPU，未覆盖旧证据。

| 更新后节点 | fit token准确率1e-4 / 5e-4 | probe token准确率1e-4 / 5e-4 | probe运动MAE 1e-4 / 5e-4 | probe归一化L1 1e-4 / 5e-4 |
|---|---|---|---|---|
| 0 | 38.39% / 38.39% | 32.14% / 32.14% | 0.100045 / 0.100045 | 0.166197 / 0.166197 |
| 10 | 97.32% / 100.00% | 32.14% / 33.93% | 0.092145 / 0.070397 | 0.162852 / 0.152872 |
| 25 | 100.00% / 100.00% | 32.14% / 28.57% | 0.087087 / 0.084092 | 0.170275 / 0.202182 |
| 50 | 100.00% / 100.00% | 32.14% / 28.57% | 0.091907 / 0.078811 | 0.177316 / 0.188476 |

两组最终fit六维运动MAE均0.001350、连续归一化L1均0.002039、夹爪16/16、近零平移0/16；离散token全对不等于连续GT零误差。最终probe夹爪1e-4为7/8、5e-4为8/8，近零命令3/8与0/8。近零阈值是命令范数<0.02，不是末端位移；运动MAE只计算六维运动、与七维归一化L1量纲不同，不合并为总体优劣。

| loss在该更新之前 | lr1e-4 | lr5e-4 |
|---|---:|---:|
| 1 | 2.819472 | 2.819472 |
| 10 | 0.283406 | 0.0000489532 |
| 25 | 0.000280078 | 0.00000622744 |
| 50 | 0.0000210585 | 0.000000450759 |

loss测量在更新前，动作指标在更新后；第0步没有单独测loss，不能把update1重新标成step0。5e-4训练阶段PyTorch峰值allocated9.219GiB（参考9.217GiB），不是系统总显存。

两组生产loader各自重载的24×7自回归动作均与训练loader一致；teacher-forcing不全等：第50步probe decoded-token L1，1e-4训练/生产0.127591/0.124930，5e-4为0.123669/0.124510。各组token accuracy训练/生产同值，但不能宣称全部logits/dtype等价。

**能得出的结论：** 此seed/固定样本中5e-4更快记忆16帧，未出现较大学习率必然无法学习。后期probe token准确率更低、归一化L1更高，但运动MAE更低、夹爪数更好；不能简单宣布整体更优或更差。训练loss持续下降不意味着跨帧指标全面提高。这与记忆/迁移不足相容，不是历史唯一根因、4bit损失或策略成功率的充分证据。8probe曾被原全量adapter训练见过，**不是held-out**；每个lr只有一个seed/run，不能声称统计显著性。

22项CPU测试通过，AST、Notebook schema及最终reader存储输出重新核验。完整报告提交Git，权重/optimizer/logs留在忽略目录、原数据和test.jpg保留；接手自动化完成后关闭，不自动扩展长训练/500评测。

下一阶段建议用户确认：从未微调LIBERO的基础权重开始，按episode隔离训练/验证，只用训练集计算动作统计，设计有界小规模验证。若仍从见过全量LIBERO的3500-step adapter开始，即使新划分episode也不能宣称从未见过的held-out。另一选择是先检查历史随机增强/裁剪影响；本轮不启动新的GPU实验。


## 35. 从 OXE 基础权重新建 LoRA：同任务完整演示隔离（启动记录，最终见第36节）

用户已授权继续。此前16帧实验从3500-step全量adapter热启动，8probe不是held-out。本轮改用原始 `openvla/openvla-7b` 缓存revision `47a0ec7fc4ec123775a391911046cf33cf9ed83f`，3个权重分片及配置/tokenizer/processor/custom code已SHA256核验，没有LoRA索引项或adapter_config；新建LoRA，不加载任何历史LIBERO adapter。不是重新训练OXE，也未重新审计OXE语料重叠。

`audit_episode_split.py` 完整扫描432条TFDS演示。同目标任务有46个候选，按完整内容指纹去重，再用seed7选择训练8条 `[185,343,79,110,72,400,75,113]` 与验证2条 `[394,212]`。完整演示指纹包含双视角编码图像、动作、state、语言与终止标记，两组不重叠。源HDF5 metadata为共享任务路径，不是唯一episode身份，也未验证原HDF5。

排除is_last或is_terminal后，训练992帧、验证246帧。动作统计只计算训练992帧；先转换夹爪为1-clip(raw,0,1)，运动六维用q01/q99归一化并clip，夹爪mask=False。验证六维超出训练区间的比例为3.66%、0%、6.50%、2.85%、0%、4.88%，不据此重算统计。这是裁剪诊断，不是模型预测误差。CPU数据证据为 `results/clean_task_split_audit_v1.json`；下面reader只读它，不启动GPU训练。

`train_clean_task.py` 独立有界试运行：NF4/BF16/doubleFalse、rank32/alpha16 all-linear、batch2×累积8、lr1e-4、seed7、无增强/无中心裁剪，最多50更新。每轮确定性打乱训练帧，50更新仅800帧，少于一轮992帧；这是新起点探针，不是足够训练或完整论文配方。

0/10/25/50节点分别记录：全部246验证帧teacher-forcing（训练仅24monitor帧）；自回归只用每条演示3个阶段帧，训练24/验证6。指标覆盖不同，不能把teacher准确率当自回归或任务成功率。最终重载实际生产loader，分开保留teacher与自回归差异。最终 `results/clean_task_v1.json` 尚待产生，无本轮闭环结果或成功率。

首次worker在GPU模型加载/更新前被指纹assert拦截：TFDS位置切片改变分片交错顺序，不等于全流episode编号。已改完整流枚举筛选，全部10条指纹重新CPU核验；按manifest顺序拼接动作，训练统计逐值一致。保留失败日志与 `retry-preupdate.json`，仅一次有界重试，不降低标准；这不是历史低成功率的已证实根因。32项CPU测试通过，不能替代本轮训练/泛化证据。

第0步真实证据 `results/clean_task_start_v1.json` 为verified_start_not_final：完整snapshot SHA/spec、base未变、真实帧身份/目标tokens和AR动作指标独立重算通过。训练24monitor token13.10%、AR运动MAE0.150306；验证246帧teacher token11.67%、loss11.527381，6个阶段帧AR运动MAE0.134617、归一化L1 0.274609、夹爪6/6。它们是未更新起点而非最终结果；全部1238帧7动作+EOS监督通过，新LoRA110,828,288可训练参数，worker已进入真实更新。

最终只读校验入口 `python -m reproduction.verify_clean_task --output reproduction/results/clean_task_verification_v1.json`：缺少最终报告即失败，不合成指标；核验四节点、完整验证teacher与固定AR身份/targets/指标、真实loss抽样及生产重载差异。已存在的证据禁止覆盖，权重/optimizer与日志留在Git忽略目录，原权重/数据/test.jpg保留。每小时低频接手：忙/无变化不通知，额度不足不绕过，后续定时运行有额度再接手，不保证刷新瞬间恢复。完成本轮核验后关闭，不自动扩大长训练/500评测。


In [1]:
import json
from pathlib import Path
project = Path.cwd() if Path('reproduction/results').exists() else Path.cwd().parent
split = json.loads((project / 'reproduction/results/clean_task_split_audit_v1.json').read_text())
assert split['status'] == 'cpu_split_verified_training_not_started'
assert all(split['checks'].values())
train, validation = split['selection']['train'], split['selection']['validation']
assert {e['content_sha256'] for e in train}.isdisjoint(e['content_sha256'] for e in validation)
assert split['training_statistics']['source'] == 'selected training episodes only'
print(f"CPU split verified: {split['dataset']['episode_count']} episodes; train {len(train)} / {split['training_statistics']['num_transitions']} frames; validation {len(validation)} / {split['validation']['num_transitions']} frames")
print('Content fingerprints disjoint; normalization from training only. No training result claimed.')


CPU split verified: 432 episodes; train 8 / 992 frames; validation 2 / 246 frames
Content fingerprints disjoint; normalization from training only. No training result claimed.


## 36. 干净起点50步完成：独立演示目标预测改善，生产重载差异需保留

真实最终报告 `results/clean_task_v1.json` 完成50次更新，原状态 **completed_loader_difference_requires_diagnosis**、production_actions_equal=False。独立核验 `clean_task_verification_v1.json` 的verified_complete只表示四节点完整SHA/spec、base未变、完整真实帧/teacher目标tokens、AR原动作/指标/摘要及50条loss/训练抽样通过，**不是生产动作全等PASS**。没有覆盖失败证据或重复训练。

| 更新后节点 | 训练24monitor teacher准确率 | 验证246帧teacher准确率 | 验证teacher loss | 验证6帧AR运动MAE | 验证6帧AR归一化L1 | 验证夹爪 |
|---|---:|---:|---:|---:|---:|---:|
| 0 | 13.10% | 11.67% | 11.527381 | 0.134617 | 0.274609 | 6/6 |
| 10 | 14.29% | 14.34% | 6.681424 | 0.161278 | 0.331879 | 5/6 |
| 25 | 24.40% | 23.23% | 3.714302 | 0.302624 | 0.579720 | 4/6 |
| 50 | 32.14% | 32.69% | 3.251718 | 0.132556 | 0.226822 | 6/6 |

表为训练准备loader（BF16 autocast）。验证246帧属于2条演示，不能当246个独立任务；训练teacher只监测24帧，不能与验证集合范围混淆。自回归仅6个固定验证阶段帧，中途误差变大，最终运动误差略降；teacher token改善不证明闭环策略提高。本轮只抽样800个不同训练帧，少于一轮992帧，不能声称收敛或完整论文复现。

更新1/10/25/50**前**实际平均训练loss为11.414975/7.326590/4.041071/2.797524；表中动作/teacher评估在更新后，第0步评估loss不是优化器训练loss。训练峰值PyTorch allocated8.773GiB，不是系统总显存。

最终实际生产loader验证teacher准确率32.75%、loss3.253109；验证6帧自回归与训练loader逐值一致，但训练24monitor有3个动作不同（185/t104、79/t0、72/t0），共27/30一致。185/t104最大六维动作差1.108100，不把所有差异称为无关紧要的小数舍入。生产训练monitor运动MAE0.136861，对应训练路径0.132429，保留各自指标，不混用。

### 有界只读重载诊断：不是检查点损坏证据

`diagnose_clean_reload.py` 仅对3个差异训练帧+3个验证控制帧推理；无optimizer/更新/rollout/checkpoint写入，证据 `clean_task_reload_diagnosis_v1.json`。相同训练准备loader重载6/6复现训练输出，生产loader6/6复现生产输出；在**同一生产实例**仅加BF16 autocast，5/6匹配训练；再应用完整k-bit准备+autocast，6/6匹配训练。snapshot保护SHA前后不变，model/config class相同。

训练准备路径的非量化参数全FP32；生产仍有2,733,760个vision和264,108,544个其他非量化参数为BF16，gradient_checkpointing也为False而训练准备为True。完整prepare同时改变dtype/flags，未做纯dtype或纯flag独立消融，不能唯一归因某层精度。此结果支持本轮加载准备差异而非保存损坏；“自回归早期token分叉放大后续差异”仅为合理机制推断，未记录logit margin。

本轮完成训练与有界诊断，**未修复生产实现、未测闭环成功率**。建议下一阶段先做显式可选的训练准备兼容推理并核验全30帧，再决定有界闭环或追加训练；需要用户选择，不自动扩大。32项CPU测试与AST/schema/reader校验收尾核验，原权重/数据/日志/test.jpg保留，收尾关闭此自动化。



In [1]:
import json
from pathlib import Path
project = Path.cwd() if Path('reproduction/results').exists() else Path.cwd().parent
result = json.loads((project / 'reproduction/results/clean_task_v1.json').read_text())
verified = json.loads((project / 'reproduction/results/clean_task_verification_v1.json').read_text())
diagnosis = json.loads((project / 'reproduction/results/clean_task_reload_diagnosis_v1.json').read_text())
assert verified['status'] == 'verified_complete'
assert verified['evaluations'] == result['evaluations']
assert result['status'] == 'completed_loader_difference_requires_diagnosis'
assert result['production_actions_equal'] is False
assert diagnosis['optimizer_updates'] == 0 and diagnosis['checkpoint_modified'] is False
assert diagnosis['prepared_reload_matches_recorded_training']
assert diagnosis['production_reload_matches_recorded_production']
assert not diagnosis['production_autocast_matches_training']
assert diagnosis['production_prepared_autocast_matches_training']
for evaluation in result['evaluations']:
    val = evaluation['splits']['validation']
    assert val['teacher_frames'] == 246 and val['autoregressive_frames'] == 6
    print(f"step {evaluation['step']}: validation teacher={val['teacher_token_accuracy']:.2%}; 6-frame AR motion MAE={val['autoregressive_motion_mae']:.6f}")
print('Integrity verified; production AR equality is False (3/24 train monitors differ, validation 6/6 equal).')
print('6-frame diagnosis: same-loader reload reproduced; autocast alone insufficient; full preparation + autocast matches.')
print('No new optimizer update, production fix, rollout or success-rate result.')


step 0: validation teacher=11.67%; 6-frame AR motion MAE=0.134617
step 10: validation teacher=14.34%; 6-frame AR motion MAE=0.161278
step 25: validation teacher=23.23%; 6-frame AR motion MAE=0.302624
step 50: validation teacher=32.69%; 6-frame AR motion MAE=0.132556
Integrity verified; production AR equality is False (3/24 train monitors differ, validation 6/6 equal).
6-frame diagnosis: same-loader reload reproduced; autocast alone insufficient; full preparation + autocast matches.
No new optimizer update, production fix, rollout or success-rate result.
